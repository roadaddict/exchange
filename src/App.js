import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { updateQuotes } from "actions/quotesActions";
import { doExchange } from "actions/walletActions";
import { Convert, Validation, ifValue } from "./helpers";
import SelectCurrency from "./components/selectCurrency";
import Rates from "./components/rates";
import WalletStatus from "./components/walletStatus";

import {
  Main,
  Separator,
  CurrencyLarge,
  CurrencyInput,
  CurrencyTo,
  Button
} from "./AppStyles";

class App extends React.Component {
  state = {
    isSourceFrom: true,
    currencyFrom: "GBP",
    currencyTo: "USD",
    valFrom: "",
    valTo: ""
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    return prevState.isSourceFrom
      ? {
          valTo: Convert(
            prevState.valFrom,
            nextProps.quotes[prevState.currencyFrom],
            nextProps.quotes[prevState.currencyTo]
          )
        }
      : {
          valFrom: Convert(
            prevState.valTo,
            nextProps.quotes[prevState.currencyTo],
            nextProps.quotes[prevState.currencyFrom]
          )
        };
  }

  componentDidMount() {
    this.props.updateQuotes();
    this.interval = setInterval(() => {
      this.props.updateQuotes();
    }, 10000);
  }

  onSelectFrom = currency => {
    this.setState({
      currencyFrom: currency
    });
  };

  onSelectTo = currency => {
    this.setState({
      currencyTo: currency
    });
  };

  changeFrom = event => {
    this.setState({
      valFrom: Validation(event.target.value),
      isSourceFrom: true
    });
  };

  changeTo = event => {
    this.setState({
      valTo: Validation(event.target.value),
      isSourceFrom: false
    });
  };

  exchange = () => {
    const { currencyFrom, currencyTo, valFrom } = this.state;
    this.isExchangeAvailable() &&
      this.props.doExchange(valFrom, currencyFrom, currencyTo);
  };

  isExchangeAvailable = () => {
    const { currencyFrom, currencyTo, valFrom, valTo } = this.state;

    return this.props.balance[currencyFrom] >= parseFloat(valFrom) &&
      currencyFrom !== currencyTo &&
      parseFloat(valFrom) > 0 &&
      parseFloat(valTo) > 0
      ? true
      : false;
  };

  render() {
    const { currencyFrom, currencyTo, valFrom, valTo } = this.state;

    return (
      <Main>
        <SelectCurrency
          selected={currencyFrom}
          onSelectCurrency={this.onSelectFrom}
        />
        <Separator>
          <div>
            <CurrencyLarge>{currencyFrom}</CurrencyLarge>
            <WalletStatus currency={currencyFrom} />
          </div>
          <div>
            <CurrencyInput
              value={valFrom}
              onChange={this.changeFrom}
              onFocus={this.changeFrom}
              onKeyPress={ifValue}
            />
            <Rates currencyFrom={currencyFrom} currencyTo={currencyTo} />
          </div>
        </Separator>

        <CurrencyTo>
          <SelectCurrency
            selected={currencyTo}
            onSelectCurrency={this.onSelectTo}
          />
          <Separator>
            <div>
              <CurrencyLarge>{currencyTo}</CurrencyLarge>
              <WalletStatus currency={currencyTo} />
            </div>
            <div>
              <CurrencyInput
                value={valTo}
                onChange={this.changeTo}
                onFocus={this.changeTo}
                onKeyPress={ifValue}
              />
              <Rates currencyFrom={currencyTo} currencyTo={currencyFrom} />
            </div>
          </Separator>
        </CurrencyTo>
        <Button
          onClick={this.exchange}
          isAvailable={this.isExchangeAvailable()}
        >
          Exchange
        </Button>
      </Main>
    );
  }
}

App.propTypes = {
  quotes: PropTypes.object.isRequired,
  updateQuotes: PropTypes.func.isRequired,
  doExchange: PropTypes.func.isRequired,
  balance: PropTypes.object.isRequired
};

export default connect(
  state => ({
    quotes: state.get("Quotes").toJS(),
    balance: state
      .get("Balance")
      .get("values")
      .toJS()
  }),
  { updateQuotes, doExchange }
)(App);
