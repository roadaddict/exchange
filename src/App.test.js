import React from 'react';
import App from './App';
import Rates from './components/rates';
import SelectCurrency from './components/selectCurrency';
import WalletStatus from './components/walletStatus';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import RootReducer from './reducers';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import { shallow } from 'enzyme';

const store = createStore(
  RootReducer,
  composeWithDevTools(applyMiddleware(thunk))
);

describe('<App /> test with Enzime', () => {
  it('renders three <Foo /> components', () => {
    const wrapper = shallow(<App />);
    // expect(wrapper.find(Foo)).to.have.lengthOf(3);
  });
});

test('General App test', () => {
  const component = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Rates component test', () => {
  const component = renderer.create(
    <Provider store={store}>
      <Rates
        currencyFrom="GBP"
        currencyTo="USD"
        quotes={{ GBP: 1, USD: 1.3 }}
      />
    </Provider>
  );
  let rates = component.toJSON();
  expect(rates).toMatchSnapshot();
});

test('Currency selector component test', () => {
  const component = renderer.create(
    <Provider store={store}>
      <SelectCurrency
        currencies={['USD', 'GBP', 'EUR']}
        selected="USD"
        onSelectCurrency={() => true}
      />
    </Provider>
  );
  let selectCurrency = component.toJSON();
  expect(selectCurrency).toMatchSnapshot();
});

test('Wallet status component test', () => {
  const component = renderer.create(
    <Provider store={store}>
      <WalletStatus
        balance={{
          values: {
            GBP: 500,
            USD: 0,
            EUR: 50
          }
        }}
        currency="EUR"
      />
    </Provider>
  );
  let walletStatus = component.toJSON();
  expect(walletStatus).toMatchSnapshot();
});
