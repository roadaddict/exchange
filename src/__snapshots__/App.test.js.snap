// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency selector component test  1`] = `
<div
  className="sc-bdVaJa imjCLz"
>
  <div
    className="sc-bwzfXH eNZAvN"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    selected={false}
    tabIndex={0}
  >
    GBP
  </div>
  <div
    className="sc-bwzfXH iFzPSC"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    selected={true}
    tabIndex={0}
  >
    USD
  </div>
  <div
    className="sc-bwzfXH eNZAvN"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    selected={false}
    tabIndex={0}
  >
    EUR
  </div>
</div>
`;

exports[`Currency selector component test 1`] = `
<div
  className="sc-bdVaJa imjCLz"
>
  <div
    className="sc-bwzfXH eNZAvN"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    selected={false}
    tabIndex={0}
  >
    GBP
  </div>
  <div
    className="sc-bwzfXH iFzPSC"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    selected={true}
    tabIndex={0}
  >
    USD
  </div>
  <div
    className="sc-bwzfXH eNZAvN"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    selected={false}
    tabIndex={0}
  >
    EUR
  </div>
</div>
`;

exports[`General App test 1`] = `
<div
  className="sc-gzVnrw fKgFaz"
>
  <div
    className="sc-bdVaJa imjCLz"
  >
    <div
      className="sc-bwzfXH iFzPSC"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      selected={true}
      tabIndex={0}
    >
      GBP
    </div>
    <div
      className="sc-bwzfXH eNZAvN"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      selected={false}
      tabIndex={0}
    >
      USD
    </div>
    <div
      className="sc-bwzfXH eNZAvN"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      selected={false}
      tabIndex={0}
    >
      EUR
    </div>
  </div>
  <div
    className="sc-htoDjs cDGZju"
  >
    <div>
      <div
        className="sc-dnqmqq gnUkkk"
      >
        GBP
      </div>
      <div
        className="sc-bZQynM hPoDmq"
      >
        <span>
          You have: 
          £
          500.00
        </span>
      </div>
    </div>
    <div>
      <input
        className="sc-iwsKbI mRYEu"
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        value=""
      />
      <div
        className="sc-htpNat eFITXY"
      >
        <div
          className="sc-bxivhb jzkpli"
        >
          <span
            className="sc-EHOje hMbVuB"
          >
            £
          </span>
          <span>
            1
          </span>
        </div>
        <div
          className="sc-ifAKCX kKKcRI"
        >
          =
        </div>
        <div
          className="sc-bxivhb jzkpli"
        >
          <span
            className="sc-EHOje hMbVuB"
          >
            $
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-gZMcBi liGWgh"
  >
    <div
      className="sc-bdVaJa imjCLz"
    >
      <div
        className="sc-bwzfXH eNZAvN"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        selected={false}
        tabIndex={0}
      >
        GBP
      </div>
      <div
        className="sc-bwzfXH iFzPSC"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        selected={true}
        tabIndex={0}
      >
        USD
      </div>
      <div
        className="sc-bwzfXH eNZAvN"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        selected={false}
        tabIndex={0}
      >
        EUR
      </div>
    </div>
    <div
      className="sc-htoDjs cDGZju"
    >
      <div>
        <div
          className="sc-dnqmqq gnUkkk"
        >
          USD
        </div>
        <div
          className="sc-bZQynM hPoDmq"
        >
          <span>
            You have: 
            $
            0
          </span>
        </div>
      </div>
      <div>
        <input
          className="sc-iwsKbI mRYEu"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          value="0.00"
        />
        <div
          className="sc-htpNat eFITXY"
        >
          <div
            className="sc-bxivhb jzkpli"
          >
            <span
              className="sc-EHOje hMbVuB"
            >
              $
            </span>
            <span>
              1
            </span>
          </div>
          <div
            className="sc-ifAKCX kKKcRI"
          >
            =
          </div>
          <div
            className="sc-bxivhb jzkpli"
          >
            <span
              className="sc-EHOje hMbVuB"
            >
              £
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="sc-gqjmRU rzgwy"
    onClick={[Function]}
  >
    Exchange
  </button>
</div>
`;

exports[`General test  1`] = `
<div
  className="sc-gzVnrw fKgFaz"
>
  <div
    className="sc-bdVaJa imjCLz"
  >
    <div
      className="sc-bwzfXH iFzPSC"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      selected={true}
      tabIndex={0}
    >
      GBP
    </div>
    <div
      className="sc-bwzfXH eNZAvN"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      selected={false}
      tabIndex={0}
    >
      USD
    </div>
    <div
      className="sc-bwzfXH eNZAvN"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      selected={false}
      tabIndex={0}
    >
      EUR
    </div>
  </div>
  <div
    className="sc-htoDjs cDGZju"
  >
    <div>
      <div
        className="sc-dnqmqq gnUkkk"
      >
        GBP
      </div>
      <div
        className="sc-bZQynM hPoDmq"
      >
        <span>
          You have: 
          £
          500.00
        </span>
      </div>
    </div>
    <div>
      <input
        className="sc-iwsKbI mRYEu"
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        value=""
      />
      <div
        className="sc-htpNat eFITXY"
      >
        <div
          className="sc-bxivhb jzkpli"
        >
          <span
            className="sc-EHOje hMbVuB"
          >
            £
          </span>
          <span>
            1
          </span>
        </div>
        <div
          className="sc-ifAKCX kKKcRI"
        >
          =
        </div>
        <div
          className="sc-bxivhb jzkpli"
        >
          <span
            className="sc-EHOje hMbVuB"
          >
            $
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-gZMcBi liGWgh"
  >
    <div
      className="sc-bdVaJa imjCLz"
    >
      <div
        className="sc-bwzfXH eNZAvN"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        selected={false}
        tabIndex={0}
      >
        GBP
      </div>
      <div
        className="sc-bwzfXH iFzPSC"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        selected={true}
        tabIndex={0}
      >
        USD
      </div>
      <div
        className="sc-bwzfXH eNZAvN"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        selected={false}
        tabIndex={0}
      >
        EUR
      </div>
    </div>
    <div
      className="sc-htoDjs cDGZju"
    >
      <div>
        <div
          className="sc-dnqmqq gnUkkk"
        >
          USD
        </div>
        <div
          className="sc-bZQynM hPoDmq"
        >
          <span>
            You have: 
            $
            0
          </span>
        </div>
      </div>
      <div>
        <input
          className="sc-iwsKbI mRYEu"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          value="0.00"
        />
        <div
          className="sc-htpNat eFITXY"
        >
          <div
            className="sc-bxivhb jzkpli"
          >
            <span
              className="sc-EHOje hMbVuB"
            >
              $
            </span>
            <span>
              1
            </span>
          </div>
          <div
            className="sc-ifAKCX kKKcRI"
          >
            =
          </div>
          <div
            className="sc-bxivhb jzkpli"
          >
            <span
              className="sc-EHOje hMbVuB"
            >
              £
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="sc-gqjmRU rzgwy"
    onClick={[Function]}
  >
    Exchange
  </button>
</div>
`;

exports[`Rates component test  1`] = `
<div
  className="sc-htpNat eFITXY"
>
  <div
    className="sc-bxivhb jzkpli"
  >
    <span
      className="sc-EHOje hMbVuB"
    >
      £
    </span>
    <span>
      1
    </span>
  </div>
  <div
    className="sc-ifAKCX kKKcRI"
  >
    =
  </div>
  <div
    className="sc-bxivhb jzkpli"
  >
    <span
      className="sc-EHOje hMbVuB"
    >
      $
    </span>
  </div>
</div>
`;

exports[`Rates component test 1`] = `
<div
  className="sc-htpNat eFITXY"
>
  <div
    className="sc-bxivhb jzkpli"
  >
    <span
      className="sc-EHOje hMbVuB"
    >
      £
    </span>
    <span>
      1
    </span>
  </div>
  <div
    className="sc-ifAKCX kKKcRI"
  >
    =
  </div>
  <div
    className="sc-bxivhb jzkpli"
  >
    <span
      className="sc-EHOje hMbVuB"
    >
      $
    </span>
  </div>
</div>
`;

exports[`Wallet status component test  1`] = `
<div
  className="sc-bZQynM hPoDmq"
>
  <span>
    You have: 
    €
    50.00
  </span>
</div>
`;

exports[`Wallet status component test 1`] = `
<div
  className="sc-bZQynM hPoDmq"
>
  <span>
    You have: 
    €
    50.00
  </span>
</div>
`;
